[FLINK-29549]- Flink Glue Catalog integration

fmorillo7694 opened the pull request:
<!--
*Thank you for contributing to Apache Flink AWS Connectors - we are happy that you want to help us improve our Flink connectors. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

## Contribution Checklist

- The name of the pull request should correspond to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
- Commits should be in the form of "[FLINK-XXXX][component] Title of the pull request", where [FLINK-XXXX] should be replaced by the actual issue number. 
    Generally, [component] should be the connector you are working on.
    For example: "[FLINK-XXXX][Connectors/Kinesis] XXXX" if you are working on the Kinesis connector or "[FLINK-XXXX][Connectors/AWS] XXXX" if you are working on components shared among all the connectors.
- Each pull request should only have one JIRA issue.
- Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.
-->

## Purpose of the change

*For example: Implements the Table API for the Kinesis Source.*

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
- *Added integration tests for end-to-end deployment*
- *Added unit tests*
- *Manually verified by running the Kinesis connector on a local Flink cluster.*

## Significant changes
*(Please check any boxes [x] if the answer is "yes". You can first publish the PR and check them afterwards, for convenience.)*
- [ ] Dependencies have been added or upgraded
- [ ] Public API has been changed (Public API is any class annotated with `@Public(Evolving)`)
- [ ] Serializers have been changed
- [ ] New feature has been introduced
  - If yes, how is this documented? (not applicable / docs / JavaDocs / not documented)

Samrat002 committed:
[FLINK-30481][FLIP-277] GlueCatalog Implementation

Co-Authored-By: Anthony Pounds-Cornish <antpc@amazon.co.uk>

fmorillo7694 committed:
AWS Glue Catalog implementation for Apache Flink

refactoring directory

Adding Parent Pom

fmorillo7694 committed:
merged commits

fmorillo7694 committed:
fixing conflicts with AWSGeneralUtil

Samrat002 commented on flink-catalog-aws/flink-catalog-aws-glue/.gitignore:
```diff
@@ -0,0 +1,2 @@
+./target
```
is this gitignore file required

Samrat002 commented on flink-catalog-aws/flink-catalog-aws-glue/.idea/.gitignore:
```diff
@@ -0,0 +1,10 @@
+# Default ignored files
```
Is this file required ?

Samrat002 commented on flink-catalog-aws/flink-catalog-aws-glue/.idea/aws.xml:
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
is this file required ?

Samrat002 commented on flink-catalog-aws/flink-catalog-aws-glue/.idea/checkstyle-idea.xml:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
why do you need this file ?

Samrat002 reviewed (changes requested):
Thanks for the contribution. A few points to note:


It looks like most of the code has already been reviewed across multiple PRs.

Concerns:

1. The current implementation exposes low-level Glue details (e.g., support for lowercase table names) directly through the Flink Catalog. Is this acceptable? I recommend raising this topic with the community. In my opinion, it would be better to encapsulate such Glue-specific behavior and avoid exposing it directly via the Flink Catalog interface. The catalog's behavior should remain consistent with other catalogs, with differences controlled via configuration only.

2. Please consider implementing the basic configuration options defined in [FLIP-277](https://cwiki.apache.org/confluence/display/FLINK/FLIP-277). If that’s not feasible in this PR, a fast follow-up would be valuable, especially for users relying on different credential modes.

3. It seems that some `.idea` folder files have been committed. Please remove them from the PR.

Cheers,
Samrat

fmorillo7694 committed:
fixing build

fmorillo7694 committed:
removing idea

fmorillo7694 committed:
removing gitignore

fmorillo7694 committed:
Adding Docs

fmorillo7694 committed:
Add AWS Glue Catalog documentation and update POM

fmorillo7694 committed:
Update AWS Glue Catalog documentation to match README

FranMorilloAWS commented:
> Thanks for the contribution. A few points to note:
> 
> It looks like most of the code has already been reviewed across multiple PRs.
> 
> Concerns:
> 
>     1. The current implementation exposes low-level Glue details (e.g., support for lowercase table names) directly through the Flink Catalog. Is this acceptable? I recommend raising this topic with the community. In my opinion, it would be better to encapsulate such Glue-specific behavior and avoid exposing it directly via the Flink Catalog interface. The catalog's behavior should remain consistent with other catalogs, with differences controlled via configuration only.
> 
>     2. Please consider implementing the basic configuration options defined in [FLIP-277](https://cwiki.apache.org/confluence/display/FLINK/FLIP-277). If that’s not feasible in this PR, a fast follow-up would be valuable, especially for users relying on different credential modes.
> 
>     3. It seems that some `.idea` folder files have been committed. Please remove them from the PR.
> 
> 
> Cheers, Samrat


Hey Samrat. How could we encapsulate this specific issues?  From user perspective we are already limiting them creating tables and databases with uppercase. In regards of the schema, we encapsulate storing the actual original columnNames in column parameters in Glue, so even though user sees their schema in glue with lower case (default for glue). we actually leverage the original column name.

2. Yes right after we get the first release delivered we will add the remainder authentication, and partitions support and alter support

3. Fixed!

leekeiabstraction commented on flink-catalog-aws/flink-catalog-aws-glue/src/main/java/org/apache/flink/table/catalog/glue/GlueCatalog.java:
```diff
@@ -0,0 +1,935 @@
+
+        try {
+            return glueFunctionsOperations.getGlueFunction(normalizedPath);
+        } catch (FunctionNotExistException e) {
+            throw e;
```
The catch and rethrow as-is is redundant. 

Do you mean to add additional information here by wrapping with another exception? Otherwise, remove the catch entirely.

nicusX commented:
@Samrat002 @FranMorilloAWS about case sensitivity, these are the logical rules I suggest:

1. the actual case should preserved for **all supported objects** and used consistently in Flink interface, e.g. in`CREATE TABLE...` but also `SHOW TABLE`, `SHOW COLUMNS ...` etc. From a user's perspective, the way object names are actually stored in Glue should not be a concern.
2. The additional metadata should be stored in Glue in a consistent way for all object types. `Parameters` seems to be available for Database, Tables, and Columns in Glue. A problem arises for Functions which don't have any `Parameters` if I am not mistaken. 
3. The additional metadata should not leak into Flink user interface. The "originalName" should not pop up in a `SHOW TABLES` or `SHOW COLUMNS`
4. The fact Glue UI is limited and only shows lowercase names is not a concern of this component's

I would add that the way the additional metadata (the original name and any additional info required) is stored in Glue should be clearly documented, in case a user want's to build their own external interface to extract or update information in the Glue Catalog.

fmorillo7694 committed:
support for uppercase for database and table name

FranMorilloAWS commented:
> Thanks for the contribution. A few points to note:
> 
> It looks like most of the code has already been reviewed across multiple PRs.
> 
> Concerns:
> 
>     1. The current implementation exposes low-level Glue details (e.g., support for lowercase table names) directly through the Flink Catalog. Is this acceptable? I recommend raising this topic with the community. In my opinion, it would be better to encapsulate such Glue-specific behavior and avoid exposing it directly via the Flink Catalog interface. The catalog's behavior should remain consistent with other catalogs, with differences controlled via configuration only.
> 
>     2. Please consider implementing the basic configuration options defined in [FLIP-277](https://cwiki.apache.org/confluence/display/FLINK/FLIP-277). If that’s not feasible in this PR, a fast follow-up would be valuable, especially for users relying on different credential modes.
> 
>     3. It seems that some `.idea` folder files have been committed. Please remove them from the PR.
> 
> 
> Cheers, Samrat

Hello @Samrat002 @nicusX. Using Database Parameters and Table Parameters we are now able to use lower/upper case for defining the Database and Table Name.  By using the Show Tables/Show Databases, Describe table commands, we will show the original Flink Definition, even though in Glue UI it will be all in lower case

fmorillo7694 committed:
adding lowercase support for database name

fmorillo7694 committed:
adding upper case support for table names

fmorillo7694 committed:
updating tests

Samrat002 commented:
Thanks for incorporating the changes. I will review the pr in next couple of days .

Samrat002 reviewed (commented):
Patch looks great. Thank you for making the change. 
minor comments 
Cheers, 
Samrat

fmorillo7694 committed:
Merge remote-tracking branch 'upstream/main' into new-combined-implementation

fmorillo7694 committed:
Update flink-catalog-aws for Flink 2.0 / 6.0-SNAPSHOT compatibility

- Remove scala.binary.version property (Flink 2.0 dropped Scala deps)
- Replace flink-table-planner_2.12 with flink-table-planner-loader
- Migrate CatalogTable.of() to CatalogTable.newBuilder() API (removed in Flink 2.0)
- All 120 tests pass

FCAgent committed:
Merge remote-tracking branch 'upstream/main' into new-combined-implementation

FCAgent committed:
[hotfix] Fix spotless formatting violations in flink-catalog-aws-glue

fmorillo7694 commented:
## Real AWS Glue integration test results (design rationale for `originalName`)

The latest head (b1a05d3) includes fixes driven by an integration test run against a **real Glue Data Catalog** (us-west-2), not Moto/Localstack. Posting the evidence here since it settles a design question a reviewer might reasonably raise: *why store lowercase names plus an `originalName` parameter instead of writing mixed-case names directly?*

### Empirical facts about real Glue (reproducible via the AWS CLI)

1. **Glue lowercases column names on `CreateTable`.** Creating a table with columns `userId` / `EventTime` stores and returns `userid` / `eventtime`. Glue does *not* preserve mixed-case column names, so writing them as-is silently loses case.
2. **Glue rejects column-level `Parameters` on partition columns** with `400: Parameters not supported for partition columns`. The `originalName` column-parameter mechanism therefore cannot be used for partition keys — their declared case is preserved in an order-preserving *table-level* parameter (`flink.original-partition-keys`) instead.

### What case loss breaks (observed before the fix, same suite)

- `DESCRIBE` returned `[userid, eventtime, username]` for a table declared `[userId, EventTime, userName]`
- `SELECT userId FROM t` → *Column 'userId' not found in any table; did you mean 'userid'?*
- Partition ops failed to match: spec `{eventDate=...}` vs stored keys `[eventdate]`
- Streaming `INSERT INTO` / `SELECT` against catalog-registered Kinesis tables failed to plan, and JSON field names on the wire would be lowercased (silent data corruption for consumers)

### Validation on the current head

25-step integration suite against real AWS, all passing:

- **DDL**: mixed-case databases, tables, and columns round-trip exactly (create / show / describe / drop); `ALTER TABLE ... SET` (now implemented via `UpdateTable`) with a case-preservation regression check; partitioned tables with `ADD/SHOW/DROP PARTITION` on mixed-case partition keys
- **Streaming**: unbounded job writing to and reading from a real Kinesis stream through source/sink tables registered in this catalog; JSON field names verified case-intact on the wire via `GetRecords`; `PARTITIONED BY` on a Kinesis table verified to set the record partition key from the (case-restored) column
- Unit tests: 126/126

One known behavior worth documenting: partition columns are appended after data columns on read-back (Glue models partition keys outside the storage descriptor, like Hive), so declaring partition columns last keeps the schema order stable across the round-trip.
